### PR TITLE
add partial denoising to Unique3D normal diffusion

### PR DESCRIPTION
**Pull request by kampelmuehler**

### CODE REVIEW

File: patches/comfyui_3d_pack.diff [hunk 1]
```diff
@@ -1,3 +1,218 @@
...
++                             generator=generator,
++                             device=device,
++                             dtype=dtype)
++
++        # get latents
++        init_latents = self.scheduler.add_noise(init_latents, noise, timestep)
++        latents = init_latents
++
++        return latents
++
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:106]: Instead of copy-pasting prepare latents, weren't that possible to do it like this (much less changes required): https://discord.com/channels/878201015424393246/1239916944275017819/1308813032054591614

File: patches/comfyui_3d_pack.diff [hunk 2]
```diff
@@ -1,3 +1,218 @@
...
++        if isinstance(generator, list):
++            init_latents = [
++                self.vae.encode(image[i:i + 1]).latent_dist.sample(
++                    generator[i]) for i in range(batch_size)
++            ]
++            init_latents = torch.cat(init_latents, dim=0)
++        else:
++            init_latents = self.vae.encode(image).latent_dist.sample(generator)
++
++        init_latents = 0.18215 * init_latents  # https://github.com/huggingface/diffusers/issues/437
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:78]: Use `latent_formats.SD15.process_in` for elegance

File: patches/comfyui_3d_pack.diff [hunk 3]
```diff
@@ -1,3 +1,218 @@
...
++        image = image.to(device=device, dtype=dtype)
++        if isinstance(generator, list) and len(generator) != batch_size:
++            raise ValueError(
++                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
++                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
++            )
++
++        if isinstance(generator, list):
++            init_latents = [
++                self.vae.encode(image[i:i + 1]).latent_dist.sample(
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:71]: separate encoding from this function and pass the latents, making the prepare_latents function minimally different from the original one

File: patches/comfyui_3d_pack.diff [hunk 4]
```diff
@@ -1,3 +1,218 @@
...
++            )
++        else:
++            init_latents = torch.cat([init_latents], dim=0)
++
++        shape = init_latents.shape
++        noise = randn_tensor(shape,
++                             generator=generator,
++                             device=device,
++                             dtype=dtype)
++
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:100]: is this piece copied from somewhere ?

File: patches/comfyui_3d_pack.diff [hunk 5]
```diff
@@ -1,3 +1,218 @@
...
++    def get_timesteps(self, num_inference_steps, new_steps, denoise):
++        # get the original timestep using init_timestep
++        init_timestep = min(int(num_inference_steps * denoise),
++                            num_inference_steps)
++        if init_timestep == 0:
++            raise ValueError(f"denoise {denoise} is too small.")
++        t_start = max(new_steps - num_inference_steps, 0)
++        timesteps = self.scheduler.timesteps[t_start:]
++
++        return timesteps, new_steps - t_start
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:31]: if it is copied unchanged, let's just call it, but pass the current "self" even if it is a different class object -- only field used is just scheduler

File: patches/comfyui_3d_pack.diff [hunk 6]
```diff
@@ -1,3 +1,131 @@
...
+-            latents,
+-        )
++        # encode reference_normals
++        reference_normals = reference_normals.permute(0, 3, 1, 2)
++        reference_normals = reference_normals.to(device=device, dtype=image_embeddings.dtype)
++        init_latents = self.vae.encode(reference_normals).latent_dist.sample(generator)
++        init_latents = self.vae.config.scaling_factor * init_latents
++        # add noise
++        noise = randn_tensor(init_latents.shape, generator=generator, device=device, dtype=init_latents.dtype)
++        latents = self.scheduler.add_noise(init_latents, noise, latent_timestep)
```
Comment (msypetkowski) [inline patches/comfyui_3d_pack.diff:92]: reference normals should be optional
